builtin/providers/ibmcloud/resource_ibmcloud_infra_ssh_key.go

```diff
@@ -0,0 +1,229 @@
+package ibmcloud
+
+import (
+	"crypto/md5"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/softlayer/softlayer-go/datatypes"
+	"github.com/softlayer/softlayer-go/filter"
+	"github.com/softlayer/softlayer-go/services"
+	"github.com/softlayer/softlayer-go/sl"
+)
+
+func resourceIBMCloudInfraSSHKey() *schema.Resource {
+	return &schema.Resource{
+		Create:   resourceIBMCloudInfraSSHKeyCreate,
+		Read:     resourceIBMCloudInfraSSHKeyRead,
+		Update:   resourceIBMCloudInfraSSHKeyUpdate,
+		Delete:   resourceIBMCloudInfraSSHKeyDelete,
+		Exists:   resourceIBMCloudInfraSSHKeyExists,
+		Importer: &schema.ResourceImporter{},
+
+		Schema: map[string]*schema.Schema{
+			"label": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"public_key": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					return strings.TrimSpace(old) == strings.TrimSpace(new)
+				},
+			},
+
+			"fingerprint": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"notes": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  nil,
+			},
+		},
+	}
+}
+
+func resourceIBMCloudInfraSSHKeyCreate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetSecuritySshKeyService(sess)
+
+	// First check if the key exists by fingerprint
+	// If so, set the Id (and fingerprint), but update notes and label (if any)
+	key := d.Get("public_key").(string)
+	label := d.Get("label").(string)
+
+	fingerprint, err := computeSSHKeyFingerprint(key)
+	if err != nil {
+		return err
+	}
+
+	keys, err := services.GetAccountService(sess).
+		Filter(filter.Path("sshKeys.fingerprint").Eq(fingerprint).Build()).
+		GetSshKeys()
+	if err == nil && len(keys) > 0 {
+		slKey := keys[0]
+		id := *slKey.Id
+		slKey.Id = nil
+		d.SetId(fmt.Sprintf("%d", id))
+		d.Set("fingerprint", fingerprint)
+		editKey := false
+
+		notes := d.Get("notes").(string)
+		if notes != "" && (slKey.Notes == nil || notes != *slKey.Notes) {
+			slKey.Notes = sl.String(notes)
+			editKey = true
+		} else if slKey.Notes != nil {
+			d.Set("notes", *slKey.Notes)
+		}
+
+		if label != *slKey.Label {
+			slKey.Label = sl.String(label)
+			editKey = true
+		}
+
+		if editKey {
+			_, err = service.Id(id).EditObject(&slKey)
+			return err
+		}
+
+		return nil
+	} // End of "Import"
```

Comment:
Right. The only reason to keep it here is to deal with cases where the same key is attempted to be created twice due to some reason which will not be allowed actually and the API will simply report error that the key already exists. In such case we simply update the notes and label if any without erring.